storage/pinlogs_blockwise.h

```diff
@@ -0,0 +1,76 @@
+
+// The length of the PIN entry log or the PIN success log in words.
+#define PIN_LOG_WORDS 2
+
+// The length of a word in bytes.
+#define WORD_SIZE (sizeof(uint32_t))
+
+uint16_t crc16(const uint8_t *data, uint32_t length) {
+  //  poly=0x1021, init=0xFFFF, xorout=0x0000 = CRC-16/CCITT-FALSE
+  uint8_t x;
+  uint16_t crc = 0xFFFF;
+
+  while (length--) {
+    x = crc >> 8 ^ *data++;
+    x ^= x >> 4;
+    crc = (crc << 8) ^ ((uint16_t)(x << 12)) ^ ((uint16_t)(x << 5)) ^
+          ((uint16_t)x);
+  }
+  return crc;
+}
+
+static secbool pin_get_fails(uint32_t *ctr) {
+  const void *logs = NULL;
+  uint16_t len = 0;
+
+  wait_random();
+
+  if (sectrue != norcow_get(PIN_LOGS_KEY, &logs, &len) ||
+      len != PIN_LOG_WORDS * WORD_SIZE) {
+    handle_fault("no PIN logs");
+    return secfalse;
+  }
+
+  if (((const uint16_t *)logs)[0] != (uint16_t) ~((const uint16_t *)logs)[1]) {
+    handle_fault("PIN logs corrupted");
+    return secfalse;
+  }
+
+  uint16_t crc = crc16(logs, len - 2);
+  if (crc != ((const uint16_t *)logs)[3]) {
+    handle_fault("PIN logs corrupted");
+    return secfalse;
+  }
+
+  *ctr = ((const uint16_t *)logs)[2];
+
+  return sectrue;
+}
+
+static secbool pin_logs_init(uint32_t fails) {
+  wait_random();
+
+  uint32_t key = random_uniform(0xFFFF);
+
+  uint16_t logs[4] = {0};
+  logs[0] = key;
+  logs[1] = ~key;
```

Comment:
changing it in https://github.com/trezor/trezor-firmware/commit/e280c472a3ed90e880bcb81bc017d2e3a18b14a2 
- assuming `random_uniform(0xFFFFFFFF)` was meant to be used in key generation
- the proposed key generation gives 0b01 for 0b1 (instead of outlined 0x10), assuming its ok
- the proposed check_key function returns secbool, but assumes sectrue==0xaaaaaaaa which seems suboptimal. I have changed it to `return ((key ^ (key << 1)) & sectrue);` but for some other sectrue values the check might not actually be sufficient like this. Maybe `return (((key ^ (key << 1)) & 0xAAAAAAAA) == 0xAAAAAAAA) ? sectrue : secfalse;` would be the way to go here?

Reply:
> assuming random_uniform(0xFFFFFFFF) was meant to be used in key generation

Yes, good catch.

> the proposed key generation gives 0b01 for 0b1 (instead of outlined 0x10), assuming its ok

True.

> the proposed check_key function returns secbool, but assumes sectrue==0xaaaaaaaa which seems suboptimal. I have changed it to return ((key ^ (key << 1)) & sectrue); but for some other sectrue values the check might not actually be sufficient like this. Maybe return (((key ^ (key << 1)) & 0xAAAAAAAA) == 0xAAAAAAAA) ? sectrue : secfalse; would be the way to go here?

I understand. We can simplify by adding `static_assert(sectrue = 0xAAAAAAAA);` so that we are alerted if the constant changes. The option, `return (((key ^ (key << 1)) & 0xAAAAAAAA) == 0xAAAAAAAA) ? sectrue : secfalse;` is ok as well. Initially, I wanted to avoid using conditional branching because it seemed more prone to fault injection attacks. However, I now think this isn't a good reason as conditional branching generally cannot be avoided.

Reply:
using multiplication to avoid conditional: https://github.com/trezor/trezor-firmware/commit/2f1e91e1d738ef021f97b9eac6834db6eea2699c

Reply:
It looks alright.